### PR TITLE
8318302: ThreadCountLimit.java failed with "Native memory allocation (mprotect) failed to protect 16384 bytes for memory to guard stack pages"

### PR DESCRIPTION
**Pull request by dholmes-ora**
The test tries to create as many threads as possible in 5 seconds, with the intention of hitting platform limits on the maximum number of threads that can be created, so that we see that the VM reacts in a reasonable way. However on Linux, if the thread creation limit is high enough, and things run fast enough, then we can instead hit a fatal error when a mmap/mprotect returns E_NOMEM because the maximum number of memory mappings (default 65530) has been exceeded.

The fix to the test (courtesy @stefank ) is to re-exec the test on Linux with `ulimit -u` set to a much smaller value than the default (I chose 4096 as it still showed over 2000 threads were created before failure). This means we are unlikely to hit the default memory mapping limit before we hit the max threads limit.

I also updated the VM's OOM error message to include exceeding the maximum number of memory mappings as a possible cause (with a hint as to what to look for in the hs_err log).

Testing:
 - ran the test 10x on all platforms in our CI (plus local testing)

Thanks

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318302](https://bugs.openjdk.org/browse/JDK-8318302): ThreadCountLimit.java failed with "Native memory allocation (mprotect) failed to protect 16384 bytes for memory to guard stack pages" (**Bug** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [b0ebf923](https://git.openjdk.org/jdk/pull/17959/files/b0ebf923bc2a6377100a9ccc182371bbac085a6d)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [263c7c41](https://git.openjdk.org/jdk/pull/17959/files/263c7c41815c40ee8e4b99f3b02f1ddf7253d046)

### Contributors
 * Stefan Karlsson `<stefank@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17959/head:pull/17959` \
`$ git checkout pull/17959`

Update a local copy of the PR: \
`$ git checkout pull/17959` \
`$ git pull https://git.openjdk.org/jdk.git pull/17959/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17959`

View PR using the GUI difftool: \
`$ git pr show -t 17959`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17959.diff">https://git.openjdk.org/jdk/pull/17959.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17959#issuecomment-1958869358)

### CODE REVIEW

**Comment by shipilev** [comment]
Maybe deserves a little synopsis change: "8318302: ThreadCountLimit.java failed due to too many memory mappings"

**Review by stefank — APPROVED** [review]
Thanks for handling this issue.

File: test/hotspot/jtreg/runtime/Thread/ThreadCountLimit.java [hunk 1]
```diff
@@ -113,13 +139,12 @@ public static void main(String[] args) {
 
     // Now that all threads have joined, we are away from dangerous
     // VM state and have enough memory to perform any other things.
-    if (reachedTimeLimit) {
-       // Windows path or a system with very large ulimit
+    if (!reachedNativeOOM) {
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/Thread/ThreadCountLimit.java:142]: Okay. It was an arbitrary choice as to which flag I kept.

File: test/hotspot/jtreg/runtime/Thread/ThreadCountLimit.java [hunk 2]
```diff
@@ -61,21 +67,42 @@ public void run() {
...
+      if (Platform.isLinux()) {
+        // On Linux this test sometimes hits the limit for the maximum number of memory mappings,
+        // which leads to various other failure modes. Run this test with a limit on how many
+        // threads the process is allowed to create, so we hit that limit first.
+
+        final String ULIMIT_CMD = "ulimit -u 4096";
+        ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder(ThreadCountLimit.class.getName());
+        String java_cmd = ProcessTools.getCommandLine(pb);
+        // Relaunch the test with args.length > 0, and the ulimit set
+        ProcessTools.executeCommand("bash", "-c", ULIMIT_CMD + " && " + java_cmd + " dummy")
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/Thread/ThreadCountLimit.java:82]: Doesn't that presume to know where bash is installed? And if it isn't found then should we skip the test due to the risk of failure?

**Comment by dholmes-ora** [comment]
> Maybe deserves a little synopsis change: "8318302: ThreadCountLimit.java failed due to too many memory mappings"

I don't object to people renaming issues to reflect the actual cause of a bug but I also don't advocate it. I spend (too much) time mapping failures to bug reports and it is much easier when the bug title reflect the failure mode rather than the bug cause. So I will keep it as is thanks.

**Comment by dholmes-ora** [comment]
@shipilev , @stefank  thank you for the reviews, and apologies for the delay in getting back to this.

**Comment by tstuefe** [comment]
So... if we reach the limit "max processes per user" we expect OOM, but if we reach vm.max_map_count, a fatal error is okay?

**Comment by dholmes-ora** [comment]
> So... if we reach the limit "max processes per user" we expect OOM, but if we reach vm.max_map_count, a fatal error is okay?

Yes. We can recover from `max processes per user` being reached, but we can't always recover from the max mappings being reached (there is actually just one fatal case when a new thread creates its own stack guard pages - there is no way to report a failure of that, so it is a fatal error).

In any case this PR is just about preventing this test from hitting a different limit to what it is trying to check.

**Review by stefank — APPROVED** [review]

**Comment by tstuefe** [comment]
> > So... if we reach the limit "max processes per user" we expect OOM, but if we reach vm.max_map_count, a fatal error is okay?
> 
> Yes. We can recover from `max processes per user` being reached, but we can't always recover from the max mappings being reached (there is actually just one fatal case when a new thread creates its own stack guard pages - there is no way to report a failure of that, so it is a fatal error).
> 
> In any case this PR is just about preventing this test from hitting a different limit to what it is trying to check.

Thank you for the clarification. The fix seems good.

File: test/hotspot/jtreg/runtime/Thread/ThreadCountLimit.java [hunk 3]
```diff
@@ -61,21 +67,42 @@ public void run() {
...
+    if (args.length == 0) {
+      // Called from the driver process so exec a new JVM on Linux.
+      if (Platform.isLinux()) {
+        // On Linux this test sometimes hits the limit for the maximum number of memory mappings,
+        // which leads to various other failure modes. Run this test with a limit on how many
+        // threads the process is allowed to create, so we hit that limit first.
+
+        final String ULIMIT_CMD = "ulimit -u 4096";
+        ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder(ThreadCountLimit.class.getName());
+        String java_cmd = ProcessTools.getCommandLine(pb);
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/Thread/ThreadCountLimit.java]: Fixed.

File: test/hotspot/jtreg/runtime/Thread/ThreadCountLimit.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/Thread/ThreadCountLimit.java]: @shipilev do you want this explored further?

[the remark this answers is not in the document]
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/Thread/ThreadCountLimit.java]: Okay thanks for the review!

**Pull request closed without merging**